instrumentation/excon/lib/opentelemetry/instrumentation/excon/middlewares/tracer_middleware.rb

```diff
@@ -22,23 +22,27 @@ class TracerMiddleware < ::Excon::Middleware::Base
           end.freeze
 
           def request_call(datum)
-            begin
-              unless datum.key?(:otel_span)
-                http_method = HTTP_METHODS_TO_UPPERCASE[datum[:method]]
-                attributes = span_creation_attributes(datum, http_method)
-                tracer.start_span(
-                  HTTP_METHODS_TO_SPAN_NAMES[http_method],
-                  attributes: attributes,
-                  kind: :client
-                ).tap do |span|
-                  datum[:otel_span] = span
-                  OpenTelemetry::Trace.with_span(span) do
-                    OpenTelemetry.propagation.inject(datum[:headers])
-                  end
-                end
-              end
-            rescue StandardError => e
-              OpenTelemetry.logger.debug(e.message)
+            return @stack.request_call(datum) if skip_trace?(datum)
+
+            http_method = HTTP_METHODS_TO_UPPERCASE[datum[:method]]
+
+            attributes = {
+              OpenTelemetry::SemanticConventions::Trace::HTTP_METHOD => http_method,
+              OpenTelemetry::SemanticConventions::Trace::HTTP_SCHEME => datum[:scheme],
+              OpenTelemetry::SemanticConventions::Trace::HTTP_TARGET => datum[:path],
+              OpenTelemetry::SemanticConventions::Trace::HTTP_HOST => datum[:host],
+              OpenTelemetry::SemanticConventions::Trace::NET_PEER_NAME => datum[:hostname],
+              OpenTelemetry::SemanticConventions::Trace::NET_PEER_PORT => datum[:port]
+            }
+
+            peer_service = Excon::Instrumentation.instance.config[:peer_service]
+            attributes[OpenTelemetry::SemanticConventions::Trace::PEER_SERVICE] = peer_service if peer_service
+            attributes.merge!(OpenTelemetry::Common::HTTP::ClientContext.attributes)
+
+            datum[:otel_span] = tracer.start_span(HTTP_METHODS_TO_SPAN_NAMES[http_method], attributes: attributes, kind: :client)
+
+            OpenTelemetry::Trace.with_span(datum[:otel_span]) do
```

Comment:
This is not related to your PR, and I am not familiar with Excon, but I think this is not implemented the way I expect. 

I would have expected `@stack.request_call(datam)` to be nested under `with_span` to ensure that any other middlewares would also have access to the `current_span` and the correct context. 

This instrumentation is sharing the span in the datum but then _disconnecting_ the context it belongs to. You do not have to address this as part of the PR but I imagine this _may_ cause problems for us when we use it in our applications.

Reply:
I went ahead and did that. However, I would have expected the current span to be in context until another span took over or until the span was finished. Is there a way to start the span, let it be in context, and then finish it without requiring the code to be run in a block?

For example, the span should be in context until we finish it in the response or error call methods.

Reply:
Hmmm. You know what, I am second guessing now since I am not sure what the underlying behavior of the stack is. 

I think you are right and we probably do not want to use `with_span` but rather manage the context stack ourselves. 

Checkout how we do that in these instrumentations that use callback hooks and cannot be wrapped in a block:

https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/main/instrumentation/racecar/lib/opentelemetry/instrumentation/racecar/process_message_subscriber.rb#L21

https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/main/instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/event_handler.rb#L57

Reply:
I went ahead and switch to calls to `OpenTelemetry::Context.attach` and `OpenTelemetry::Context.detach` based on [HTTPX](https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/02799c87f9bc9fb8af4ee90b2d704f1b213c891f/instrumentation/httpx/lib/opentelemetry/instrumentation/httpx/plugin.rb#L60) and other examples in the repo. The links you shared do the same thing, but it was a bit harder for me to follow what exactly they were doing.